### PR TITLE
Add direct copy fast path for portable copy op

### PR DESCRIPTION
**Pull request by GregoryComer**
Summary:
The PR adds a direct memcpy fast-path for portable copy and copy_ ops. This speeds up copy significantly in cases where no broadcasting is needed. Note that the copy op always checks that dim order and dtype match, so this should be sound in all cases where the shape matches (no broadcasting).

This is most noticeable when copying buffer mutations back, such as transformer KV cache when managing the cache as a mutable buffer. Prior to this change, an encoder/decoder model was taking ~25% of the total runtime copying KV cache back after permuting. With this change, the copy becomes significantly cheaper.

I benchmarked a simple model on S23 and Pixel 5:
```
class TestModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("buffer", torch.zeros((2, 10, 1024, 1024)))

    def forward(self, x):
        self.buffer.add_(x)
        return self.buffer


model = TestModel()
inputs = (torch.randn(2, 10, 1024, 1024),)

lowered = to_edge_transform_and_lower(
    torch.export.export(model, inputs),
    partitioner=[XnnpackPartitioner()],
).to_executorch()
```

S23, average of 50 runs, time in copy_:
4.1ms vs 22.3ms

Pixel 5, average of 50 runs, time in copy_:
12.1ms vs 66.6ms

This is approximately a ~5.5x speedup of the copy operator.

Differential Revision: D73656456

### CODE REVIEW

File: kernels/portable/cpu/op_copy.cpp [hunk 1]
```diff
@@ -46,17 +46,23 @@ Tensor& copy_out(
...
-        in,
-        utils::SupportedTensorDtypes::REALHBBF16,
-        src,
-        utils::SupportedTensorDtypes::REALHBBF16,
-        out,
-        utils::SupportedTensorDtypes::REALHBBF16);
-  });
+  // Use direct copy fast path if broadcast is not needed and tensors are
+  // non-empty
+  if (tensors_have_same_shape(out, src) && src.numel() > 0) {
```
Comment (swolchok) [inline kernels/portable/cpu/op_copy.cpp:51]: AIUI the requirement is looser than this; leading 1s can be ignored. how about using sizes_match_ignoring_leading_1s?

File: kernels/portable/cpu/op_copy.cpp [hunk 2]
```diff
@@ -79,17 +85,23 @@ Tensor& copy_(
...
-        in,
-        utils::SupportedTensorDtypes::REALHBBF16,
-        src,
-        utils::SupportedTensorDtypes::REALHBBF16,
-        in,
-        utils::SupportedTensorDtypes::REALHBBF16);
-  });
+  // Use direct copy fast path if broadcast is not needed and tensors are
+  // non-empty
+  if (tensors_have_same_shape(in, src) && src.numel() > 0) {
```
Comment (swolchok) [inline kernels/portable/cpu/op_copy.cpp:90]: ditto

**Review by swolchok — APPROVED** [review]

**Comment by GregoryComer** [comment]
Overriding lint failure for land - broken trunk.
